Sources/Basics/AuthorizationProvider.swift

```diff
@@ -24,3 +42,224 @@ extension AuthorizationProvider {
+                callback(.failure(error))
+            }
+        } catch {
+            callback(.failure(error))
+        }
+    }
+    
+    public func authentication(for url: Foundation.URL) -> (user: String, password: String)? {
+        guard let server = self.server(for: url) else {
+            return nil
+        }
+        
+        do {
+            guard let existingItem = try self.search(server: server, protocol: self.`protocol`(for: url)) as? [String : Any],
+                  let passwordData = existingItem[kSecValueData as String] as? Data,
+                  let password = String(data: passwordData, encoding: String.Encoding.utf8),
+                  let account = existingItem[kSecAttrAccount as String] as? String else {
+                throw AuthorizationProviderError.unexpectedError("Failed to extract credentials for server \(server) from keychain")
+            }
+            return (user: account, password: password)
+        } catch {
+            switch error {
+            case AuthorizationProviderError.notFound:
+                ObservabilitySystem.topScope.emit(info: "No credentials found for server \(server) in keychain")
+            case AuthorizationProviderError.unexpectedError(let detail):
+                ObservabilitySystem.topScope.emit(error: detail)
+            default:
+                ObservabilitySystem.topScope.emit(error: "Failed to find credentials for server \(server) in keychain: \(error)")
+            }
+            return nil
+        }
+    }
+    
+    private func create(server: String, `protocol`: CFString, account: String, password: Data) throws {
+        let query: [String: Any] = [kSecClass as String: kSecClassInternetPassword,
+                                    kSecAttrServer as String: server,
+                                    kSecAttrProtocol as String: `protocol`,
+                                    kSecAttrAccount as String: account,
+                                    kSecValueData as String: password]
+        
+        let status = SecItemAdd(query as CFDictionary, nil)
+        guard status == errSecSuccess else {
+            throw AuthorizationProviderError.unexpectedError("Failed to save credentials for server \(server) to keychain: status \(status)")
+        }
+    }
+    
+    private func update(server: String, `protocol`: CFString, account: String, password: Data) throws {
+        let query: [String: Any] = [kSecClass as String: kSecClassInternetPassword,
+                                    kSecAttrServer as String: server,
+                                    kSecAttrProtocol as String: `protocol`]
+        let attributes: [String: Any] = [kSecAttrAccount as String: account,
+                                         kSecValueData as String: password]
+        
+        let status = SecItemUpdate(query as CFDictionary, attributes as CFDictionary)
+        guard status != errSecItemNotFound else {
+            throw AuthorizationProviderError.notFound
```

Comment:
Instead of having a semi-private `notFound` error case, this method could return a Boolean.